### PR TITLE
Add CMake support

### PR DESCRIPTION
**Pull request by oold**
This will allow for a smoother initialization on Win32 by doing that work on DLL load. I've also taken this opportunity to remove the special, public, Win32-specific functions and I've reworked the list structures. The VS files were removed as they can now be generated via CMake.

Bonus: The CMake support can also be used for other targets and will simply add the `include` directory.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -34,12 +34,16 @@ If you wish to use the pthreads implementation on Windows, in preference to the
...
 
 ### With Win32 threads
-To use C11 threads over the Windows threads API, beyond adding `c11threads.h` to
-your project, you also need to compile `c11threads_win32.c` as part of your
-build. Additionally, if you're in a situation where letting c11threads keep
-resources for the duration of the process lifetime is not desirable, you can
-call `c11threads_destroy_win32()` to free them manually at any point, when
-you're done with it.
+
+Simply add the CMake project to your own project or build a shared library from
```
Comment (jtsiomb) [inline README.md]: I don't like the cmake-specific wording. I prefer instructing people to compile `c11threads_win32.c`, and leave them to do it with whatever build system they choose to.

If there is some valueable advice for cmake users, it should be less prominent, in the form of "if you use cmake, you can do X Y and Z", but if it's not some arcane knowledge we're talking about, and it's plain enough what needs to be done, I'd rather leave it at the build-system agnostic instructions.

File: README.md [hunk 2]
```diff
@@ -34,12 +34,16 @@ If you wish to use the pthreads implementation on Windows, in preference to the
...
+`c11threads_win32.c`. Threads need to be created with `thrd_create()` to work
+properly with `thrd_join()` and `thrd_detach()`.
+
```
Comment (jtsiomb) [inline README.md]: I'm glad we no longer need special destroy instructions for win32. But is it really necessary to spell out that threads need to be created with thrd_create? Surely nobody would expect they should call CreateThread() and then continue with thrd_join afterwards...

File: README.md [hunk 3]
```diff
@@ -34,12 +34,16 @@ If you wish to use the pthreads implementation on Windows, in preference to the
...
+```cmake
+add_subdirectory(path_to_c11threads)
+add_executable(my_program program.c)
+target_link_libraries(my_program c11threads)
+```
```
Comment (jtsiomb) [inline README.md]: If this is non-obvious to cmake users, I'd like it in its own subsection... but nevermind I'll reformat this myself.

**Review by jtsiomb — CHANGES_REQUESTED** [review]
Unless I'm misunderstanding, there are some code improvements, AND the addition of a cmake build file in this pull request. I want the additions, while I'm not decided on cmake. Please split this into two separate pull requests.

**Comment by oold** [comment]
I'll submit the list changes as a separate PR.

**Comment by jtsiomb** [comment]
I applied the other pull requests, so this needs rebasing to only contain the cmake-specific changes. However I was thinking about how I want to deal with build systems, and I think I want to make them less prominent. I want to emphasize that people can just grab `c11theads.h` (and the source file for windows) and drop it in their projects, and to not be put off by seing an unfamiliar build file and thinking they have to use it.

So what I think I want to do is to move the test out of the way in its own subdirectory, and move the Makefile, plus any other build system you want to add for windows into that directory. That makes it more obvious that the build files are only there to build the test program, and are not required for using the library itself.

From your previous changes in the readme file I take it you meant this not just for building the test program, but as a way of using c11threads in programs which use cmake themselves. I don't like that approach for such a simple library. It feels overcomplicated.

I'll make the change moving test and its build files out of the way later today. If you think there is value having cmake just to build the test program, then please wait a bit, and rebase on top of my change. Ideally I want to see a pull request with just the minimum changes necessary for the cmake build.

**Comment by oold** [comment]
We can do that, but the CMake solution makes it much easier to build a shared library from the .c file. The shared library is necessary to perform the initialization and deinitialization automatically. I can tell potential users to either somehow figure out how to build a DLL for Windows or to just go with the existing CMake files if that’s alright.

**Comment by jtsiomb** [comment]
I don't want to change this into a standalone separately compiled dll. It's overkill. This is supposed to be a simple "drop it into your code and use it" thing. If someone wants to make a DLL out of it fine, but the supported use case is the drop-in one.

**Comment by jtsiomb** [comment]
I replaced the visual studio project with an nmake makefile so that it'll work with all visual studio versions, and I think I'll leave it at that for the windows build. I'd rather not add cmake. I'll probably just add a separate mingw makefile at some point too.

**Pull request closed without merging**